src/main/rx/rx.c

```diff
@@ -533,41 +488,39 @@ static void detectAndApplySignalLossBehaviour(timeUs_t currentTimeUs)
     debug[2] = rxRuntimeConfig.rcReadRawFn(&rxRuntimeConfig, 0);
 #endif
 
-    rxResetFlightChannelStatus();
-
-    for (int channel = 0; channel < getRxChannelCount(); channel++) {
-
-        uint16_t sample = (useValueFromRx) ? rcRaw[channel] : PPM_RCVR_TIMEOUT;
+    rxFlightChannelsValid = true;
+    for (int channel = 0; channel < rxChannelCount; channel++) {
+        uint16_t sample = rcRaw[channel];
 
-        bool validPulse = isPulseValid(sample);
+        const bool validPulse = useValueFromRx && isPulseValid(sample);
 
-        if (!validPulse) {
+        if (validPulse) {
+            rcInvalidPulsPeriod[channel] = currentMilliTime + MAX_INVALID_PULS_TIME;
+        } else {
             if (currentMilliTime < rcInvalidPulsPeriod[channel]) {
                 sample = rcData[channel];           // hold channel for MAX_INVALID_PULS_TIME
             } else {
                 sample = getRxfailValue(channel);   // after that apply rxfail value
-                rxUpdateFlightChannelStatus(channel, validPulse);
+                if (channel < NON_AUX_CHANNEL_COUNT) {
+                    rxFlightChannelsValid = false;
+                }
             }
-        } else {
-            rcInvalidPulsPeriod[channel] = currentMilliTime + MAX_INVALID_PULS_TIME;
         }
-
-        if (rxIsDataDriven) {
-            rcData[channel] = sample;
-        } else {
-            rcData[channel] = calculateNonDataDrivenChannel(channel, sample);
+        rcData[channel] = sample;
+    }
+    if (feature(FEATURE_RX_PARALLEL_PWM | FEATURE_RX_PPM)) {
+        // smooth output for PWM and PPM
+        for (int channel = 0; channel < rxChannelCount; channel++) {
+            rcData[channel] = calculateChannelMovingAverage(channel, rcData[channel]);
         }
     }
 
-    rxFlightChannelsValid = rxHaveValidFlightChannels();
-
-    if ((rxFlightChannelsValid) && !IS_RC_MODE_ACTIVE(BOXFAILSAFE)) {
+    if (rxFlightChannelsValid && !IS_RC_MODE_ACTIVE(BOXFAILSAFE)) {
         failsafeOnValidDataReceived();
     } else {
-        rxIsInFailsafeMode = rxIsInFailsafeModeNotDataDriven = true;
+        rxIsInFailsafeMode = true;
```

Comment:
ad rxFlightChannelsValid:
It may say valid at most for `rcInvalidPulsPeriod` from setting rxIsInFailsafeMode, then !validPulse (from rxIsInFailsafeMode) will clear it.
And something must have set rxIsInFailsafeMode, so one of the timers will be expired anyway (BOXFAILSAFE is exception)
So in most cases `rxFlightChannelsValid = !rxIsInFailsafeMode && isPulseValid(...);`

The logic in rx.c is terrible ...